### PR TITLE
gmrender-resurrect: Add gmediarender service

### PR DESCRIPTION
**Pull request by mdorman**
###### Description of changes

This creates a systemd unit that will start and supervise the gmediarender daemon.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/audio/gmediarender.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      description = mdDoc ''
+        A UUID for uniquely identifying the endpoint.  If you have
+        multiple renderers on your network, you MUST set this.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd = {
+      packages = [ pkgs.gmrender-resurrect ];
```
[the remark this answers is not in the document]
Comment (mdorman) [reply nixos/modules/services/audio/gmediarender.nix:69]: Good point about using `cfg.package`.  I'm afraid I don't understand your question about putting the package in the `path`, though?

**Comment by mdorman** [comment]
Thanks, @pennae, for the careful review and detailed feedback.  I've resolved everything except the one comment you made that I didn't entirely understand about putting something in the `path`.

File: nixos/modules/services/audio/gmediarender.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdorman) [reply nixos/modules/services/audio/gmediarender.nix]: Ah, I understand now, and I believe you are correct, there is no actual need for this to be there.

File: nixos/modules/services/audio/gmediarender.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
+{ pkgs, lib, config, ... }:
+
+with lib;
+
+let
+  cfg = config.services.gmediarender;
+in
+{
+  options.services.gmediarender = {
+    enable = mkEnableOption (mdDoc "Enable gmediarender DLNA renderer service");
```
[the remark this answers is not in the document]
Comment (mdorman) [reply nixos/modules/services/audio/gmediarender.nix:10]: Sorry, when applying this I was paying more attention to the `mdDoc` that caused the problem with the manual build than the wording change (even though i noticed it and figured that was why you made it).

**Comment by mdorman** [comment]
Had to resolve some fallout from changing to the dynamic user, but I believe this has resolved everything mentioned.

**Review by pennae — COMMENTED** [review]
looking pretty good! could you also add a note to the changelog, and maybe a test? not sure how useful a test would be since this involves audio, so that idea might be no good.

File: nixos/modules/services/audio/gmediarender.nix [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            "${cfg.package}/bin/gmediarender " +
+            optionalString (cfg.audioDevice != null) ("--gstout-audiodevice=${utils.escapeSystemdExecArg cfg.audioDevice} ") +
+            optionalString (cfg.audioSink != null) ("--gstout-audiosink=${utils.escapeSystemdExecArg cfg.audioSink} ") +
+            optionalString (cfg.friendlyName != null) ("--friendly-name=${utils.escapeSystemdExecArg cfg.friendlyName} ") +
+            optionalString (cfg.initialVolume != 0) ("--initial-volume=${toString cfg.initialVolume} ") +
+            optionalString (cfg.port != null) ("--port=${toString cfg.port} ") +
+            optionalString (cfg.uuid != null) ("--uuid=${utils.escapeSystemdExecArg cfg.uuid} ");
+          Restart = "always";
+          RuntimeDirectory = "gmediarender";
+          Environment = "XDG_CACHE_HOME=%t/gmediarender";
```
[the remark this answers is not in the document]
Comment (mdorman) [reply nixos/modules/services/audio/gmediarender.nix:85]: Yeah, it doesn't really have any persistent state.

**Comment by mdorman** [comment]
Added to the changelog.

With regard to a test---this is a stateless conduit from network data to audio output; I just don't know how much there is to really test there.

**Comment by mdorman** [comment]
Substantial hardening done

```
$ systemd-analyze --quiet security gmediarender | grep '✗'
✗ SystemCallFilter=~@resources                                System call allow list defined for service, and @resources is included (e.g. ioprio_set is allowed)      0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                                            0.1
✗ SupplementaryGroups=                                        Service runs with supplementary groups                                                                   0.1
✗ PrivateDevices=                                             Service potentially has access to hardware devices                                                       0.2
✗ RestrictAddressFamilies=~AF_PACKET                          Service may allocate packet sockets                                                                      0.2
✗ RestrictAddressFamilies=~AF_NETLINK                         Service may allocate netlink sockets                                                                     0.1
✗ RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                                                       0.1
✗ RestrictAddressFamilies=~…                                  Service may allocate exotic sockets                                                                      0.3
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                                                    0.3
✗ DeviceAllow=                                                Service has no device ACL                                                                                0.2
✗ PrivateNetwork=                                             Service has access to the host's network                                                                 0.5
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                                         0.2
```

**Pull request merged**